namer/generatorvalidator.go

```diff
@@ -0,0 +1,145 @@
+package namer
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/tarantool/go-storage/crypto"
+	"github.com/tarantool/go-storage/hasher"
+	"github.com/tarantool/go-storage/kv"
+	"github.com/tarantool/go-storage/marshaller"
+)
+
+// Generator generates signer K/V pairs.
+// Implementation should use `generic` and will used for strong typing of the solution.
+type Generator[T any] interface {
+	Generate(name string, value T) ([]kv.KeyValue, error)
+}
+
+// Validator validates and build the object from K/V.
+type Validator[T any] interface {
+	Validate(pairs []kv.KeyValue) (T, error)
+}
+
+// DefaultGeneratorValidator represent default generator-validator.
+type DefaultGeneratorValidator[T any] struct {
+	Namer          Namer
+	Hasher         hasher.Hasher
+	SignerVerifier crypto.SignerVerifier
+	Marshaller     marshaller.Marshallable
+}
+
+// NewDefaultGeneratorValidator returns new object.
+func NewDefaultGeneratorValidator[T any](
+	namer Namer,
+	hasher hasher.Hasher,
+	signverifier crypto.SignerVerifier,
+	marshaller marshaller.Marshallable,
+) DefaultGeneratorValidator[T] {
+	return DefaultGeneratorValidator[T]{
+		Namer:          namer,
+		Hasher:         hasher,
+		SignerVerifier: signverifier,
+		Marshaller:     marshaller,
+	}
+}
+
+// Generate create KV pairs with value, hash and signature.
+func (gv DefaultGeneratorValidator[T]) Generate(name string, value T) ([]kv.KeyValue, error) {
+	if name == "" {
+		return []kv.KeyValue{}, ErrInvalidInput
+	}
+
+	var kvList []kv.KeyValue
+
+	blob, err := gv.Marshaller.Marshal(value)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal: %w", err)
+	}
+
+	hash, err := gv.Hasher.Hash(blob)
+	if err != nil {
+		return nil, fmt.Errorf("failed to hash: %w", err)
+	}
+
+	signature, err := gv.SignerVerifier.Sign(hash)
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign: %w", err)
+	}
+
+	names := gv.Namer.GenerateNames(name)
```

Comment:
So we generate names and then parse them with another method of the same class. At the moment it looks like we decode what we have just encoded. If it is the only planned usage of such classes I'd suggest to have a kind of single GenerateKeys method that generates keys from name instead of 2-phase transformation name->names->keys.